text/3389-manifest-lint.md

```diff
@@ -0,0 +1,398 @@
+rustc and other tool lints.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, you can configure lints through
+- `#[<level>(<lint>)]` or `#![<level>(<lint>)]`, like `#[forbid(unsafe)]`
+  - But this doesn't scale up with additional targets (benches, examples,
+    tests) or workspaces
+- On the command line, like `cargo clippy -- --forbid unsafe`
+  - This puts the burden on the caller
+- Through `RUSTFLAGS`, like `RUSTFLAGS=--forbid=unsafe cargo clippy`
+  - This puts the burden on the caller
+- In `.cargo/config.toml`'s `target.*.rustflags`
+  - This couples you to the running in specific directories and not running in
+    the right directory causes rebuilds
+  - The cargo team has previously stated that
+    [they would like to see package-specific config moved to manifests](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336/14?u=epage)
+
+We would like a solution that makes it easier to share across targets and
+packages for all tools.
+
+See also
+- [rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
+- [rust-lang/cargo#5034](https://github.com/rust-lang/cargo/issues/5034)
+- [EmbarkStudios/rust-ecosystem#59](https://github.com/EmbarkStudios/rust-ecosystem/issues/59)
+- [Proposal: Cargo Lint configuration](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A new `lints` table would be added to configure lints:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+and `cargo` would pass these along as flags to `rustc` and `clippy`.
+
+This would work with
+[RFC 2906 `workspace-deduplicate`](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#):
+```toml
+[lints]
+workspace = true
+
+[workspace.lints]
+unsafe = "forbid"
+```
+
+## Documentation Updates
+
+## The `lints` section
+
+*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*
+
+Override the default level of lints from different tools by assigning them to a new level in a
+table, for example:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+
+This is short-hand for:
+```toml
+[lints.rust]
+unsafe = { level = "forbid", priority = 0 }
+```
+
+`level` corresponds to the lint levels in `rustc`:
+- `forbid`
+- `deny`
+- `warn`
+- `allow`
+
+`priority` is a signed integer that controls which lints or lint groups override other lint groups:
+- lower (particularly negative) numbers have lower priority, being overridden
+  by higher numbers, and show up first on the command-line to tools like
+  `rustc`
+
+To know which table under `[lints]` a particular lint belongs under, it is the part before `::` in the lint
+name.  If there isn't a `::`, then the tool is `rust`.  For example a warning
+about `unsafe` would be `lints.rust.unsafe` but a lint about
```

Comment:
Is `rustc` treated as a known tool, just like Clippy? I think we could/should change that, which would address this concern: `rustc` would say: "Unknown tool `rustc`". After all, `rustc` lints only make sense, when compiling rustc itself or tools like Clippy. We could just add a `#![register_tool(rustc)]` to Clippy.

Reply:
I've at least called this out in  86932bd

Reply:
I don't see the point to have seperated lints tables for rustc and clippy (or any other lint tool). I think a single table is preferable and to have the clippy:: prefix like it is today in the atrribute inlined in the code.

```toml
[lints]
allow = [
  "dead_code", 
  "unused",
  "clippy::module_name_repetitions",
]
warn = [
  "clippy::disallowed-names",
]
```

Reply:
Part of the driving reason for separate tables is if we use the lints as keys to avoid having to quote the keys.  Either way, cargo will still need to process the lints to know which tool to send them to (at least in rustdocs case).

As for what format to use, that is a separate thread.